aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -0,0 +1,101 @@
+/*le aesd-circular-buffer.c
+ * @brief Functions and data related to a circular buffer imlementation
+ *
+ * @author Dan Walkes
+ * @date 2020-03-01
+ * @copyright Copyright (c) 2020
+ *
+ */
+
+#ifdef __KERNEL__
+#include <linux/string.h>
+#else
+#include <string.h>
+// #include <stdio.h>
+#endif
+
+#include "aesd-circular-buffer.h"
+
+
+/**
+ * @param buffer the buffer to search for corresponding offset.  Any necessary locking must be performed by caller.
+ * @param char_offset the position to search for in the buffer list, describing the zero referenced
+ *      character index if all buffer strings were concatenated end to end
+ * @param entry_offset_byte_rtn is a pointer specifying a location to store the byte of the returned aesd_buffer_entry
+ *      buffptr member corresponding to char_offset.  This value is only set when a matching char_offset is found
+ *      in aesd_buffer.
+ * @return the struct aesd_buffer_entry structure representing the position described by char_offset, or
+ * NULL if this position is not available in the buffer (not enough data is written).
+ */
+struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
+            size_t char_offset, size_t *entry_offset_byte_rtn )
+{
+    /**
+    * TODO: implement per description
+    */
+   // empty condition
+   *entry_offset_byte_rtn = 0;
+   if(buffer->full == false && (buffer->in_offs == buffer->out_offs)){
+        return NULL;
+   }
+
+    // printf("The offset is %lu\n ", char_offset);
+size_t sum_size = 0;
+int i = buffer->out_offs;
+int total_operations = 0;
+
+    
+    while(total_operations<AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED){
+
+       if(sum_size + buffer->entry[i].size > char_offset && char_offset >= sum_size){
+            *entry_offset_byte_rtn = char_offset - sum_size;
+            return &buffer->entry[i];
+
+       }
+
+        sum_size += buffer->entry[i].size; 
+        i = (i+1)%AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        total_operations++;
+    }
+
+    return NULL;
+}
+
+
+/**
+* Adds entry @param add_entry to @param buffer in the location specified in buffer->in_offs.
+* If the buffer was already full, overwrites the oldest entry and advances buffer->out_offs to the
+* new start location.
+* Any necessary locking must be handled by the caller
+* Any memory referenced in @param add_entry must be allocated by and/or must have a lifetime managed by the caller.
+*/
+void aesd_circular_buffer_add_entry(struct aesd_circular_buffer *buffer, const struct aesd_buffer_entry *add_entry)
+{
+    //starting condition
+   if(buffer->in_offs == buffer->out_offs && buffer->full == false){
```

Comment:
@pranjalgupta1901 I don't think you should need a special handler for this condition, I think you could handle everything with the else case logic.